src/nunit.analyzers/TestCaseSourceUsage/TestCaseSourceUsesStringAnalyzer.cs

```diff
@@ -192,20 +214,80 @@ private static void AnalyzeAttribute(SyntaxNodeAnalysisContext context)
+                                {
+                                    context.ReportDiagnostic(Diagnostic.Create(
+                                            mismatchInNumberOfTestMethodParameters,
+                                            syntaxNode.GetLocation(),
+                                            ">0",
+                                            methodRequiredParameters + methodOptionalParameters));
+                                }
+                            }
+                            else
+                            {
+                                if (methodRequiredParameters + methodOptionalParameters != 1)
+                                {
+                                    context.ReportDiagnostic(Diagnostic.Create(
+                                            mismatchInNumberOfTestMethodParameters,
+                                            syntaxNode.GetLocation(),
+                                            1,
+                                            methodRequiredParameters + methodOptionalParameters));
+                                }
+                                else
+                                {
+                                    IParameterSymbol testMethodParameter = testMethod.Parameters.First();
+
+                                    if (!NUnitEqualityComparerHelper.CanBeEqual(elementType, testMethodParameter.Type, context.Compilation))
+                                    {
+                                        context.ReportDiagnostic(Diagnostic.Create(
+                                                mismatchWithTestMethodParameterType,
+                                                syntaxNode.GetLocation(),
+                                                elementType,
+                                                testMethodParameter.Type,
+                                                testMethodParameter.Name));
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
             }
         }
 
-        private static void ReportIfSymbolNotIEnumerable(
+        private static ITypeSymbol? ReportIfSymbolNotIEnumerable(
             SyntaxNodeAnalysisContext context,
             SyntaxNode syntaxNode,
             ITypeSymbol typeSymbol)
         {
-            if (!typeSymbol.IsIEnumerable(out var _))
+            if (!typeSymbol.IsIEnumerable(out ITypeSymbol? elementType))
             {
                 context.ReportDiagnostic(Diagnostic.Create(
                     sourceDoesNotReturnIEnumerable,
                     syntaxNode.GetLocation(),
                     typeSymbol));
+                return null;
+            }
+            else
+            {
+                return elementType ?? context.Compilation.GetSpecialType(SpecialType.System_Object);

```

Comment:
Why do we default to `Object` here and not `null` (e.g. just removing the null coalescing) ?

Reply:
This deals with the case where the type is the non-generic `IEnumerable` which iterates over `object`. When `elementType is null` it means the type is not enumerable at all.

It might have been better if this was dealt with in the `IsEnumerable` method, but that is an existing method with several usages.  The current implementation of that method does allows distinguishing between `IEnumerable` and `IEnumerable<object>`.  That fact is used in some of the usages of the method.

Reply:
Thanks for the explanation. I think that the solution make sense 👍